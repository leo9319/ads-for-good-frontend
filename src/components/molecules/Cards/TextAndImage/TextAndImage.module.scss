@import '../../../../styles/breakpoints';
@import '../../../../styles/_spacing.scss';

.container {
  width: 348px;
  height: 464px;
  border-radius: 32px;
  padding: var(--spacing-2xl);
  box-shadow: 0px 1px 3px 0px #0000001a;
  background-color: #fff;
  overflow: hidden;

  &.cursor {
    cursor: pointer;
  }

  &.surface {
    border: 1px solid var(--Border-secondary, #eaebee);
  }

  &.size {
    &-1 {
      width: 348px;
      height: 464px;
    }

    &-2 {
      width: 160px;
      height: 214px;
    }
  }

  .textSection {
    min-width: 268px;
    min-height: 320px;
    width: 268px;
    height: 320px;
    text-align: left;
    margin-bottom: var(--spacing-md) !important;

    .heading {
      color: #17191f;
    }

    .description {
      color: #65676b;
    }
  }

  &.showImage {
    .textSection {
      min-width: 348px;
      min-height: 232px;
      width: 348px;
      // height: 232px;
    }
  }

  .icon {
    align-items: start !important;
    justify-content: start !important;

    > span[class~='icon'] {
      width: 64px !important;
      height: 64px !important;
    }
  }

  .image {
    width: 100%;
    border-bottom-left-radius: 32px;
    border-bottom-right-radius: 32px;
    object-fit: cover;
    max-height: 196px;
  }

  &.showImageOnly,
  &.showImage {
    padding: 0px !important;

    .textSection {
      padding: var(--spacing-2xl);
    }
  }

  &.showImageOnly {
    .image {
      max-height: 100% !important;
      height: 100% !important;
      margin-top: 0px !important;
      border-top-left-radius: 32px;
      border-top-right-radius: 32px;
    }
  }

  .heading {
    padding-bottom: var(--spacing-xs);
  }

  @media (max-width: $sm-max) {
    width: 160px;
    height: 214px;
    border-radius: 16px;
    padding: var(--spacing-md) !important;

    .icon {
      > span[class~='icon'] {
        width: 40px !important;
        height: 40px !important;
      }
    }

    .image {
      border-bottom-left-radius: 16px;
      border-bottom-right-radius: 16px;
    }

    .textSection {
      min-width: 128px !important;
      min-height: 142px !important;
      width: auto !important;
      margin-bottom: var(--spacing-xs) !important;
    }

    &.showImage {
      padding: 0px !important;
      margin-top: var(--spacing-xs) !important;

      .textSection {
        min-height: 75px !important;
        padding: var(--spacing-md) !important;
      }

      img {
        width: 161px;
        height: 90px;
      }
    }

    &.showImageOnly {
      .image {
        border-top-left-radius: 16px;
        border-top-right-radius: 16px;
      }
    }
  }
}
