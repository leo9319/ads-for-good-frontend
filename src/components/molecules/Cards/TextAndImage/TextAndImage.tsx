import React, { useId } from 'react';
import Text from '@radix-styles/atoms/Text';
import Heading from '@radix-styles/atoms/Heading';
import Card from '@radix-styles/atoms/Card';
import Icon from '@components/atoms/Icons';
import Image from '@components/atoms/Image';
import FallBackImg from '@assets/images/FallbackImage.svg';
import Box from '@radix-styles/atoms/Box';
import Flex from '@radix-styles/atoms/Flex';

import { classNames } from '@utils/common/classNames';
import Truncate from '@components/atoms/Truncate';

import styles from './TextAndImage.module.scss';

interface ImageProps {
  /**
   * The source of the image
   * @default ''
   * @property
   */
  src: string;
  /**
   * The alt text of the image
   * @default 'Card Image'
   * @property
   */
  alt: string;
}

export interface TextAndImageProps {
  /**
   * The variant of the card, that will change the border style of the card
   * @default default
   * @property
   */
  variant?: 'default' | 'surface';
  /**
   * The headline of the card
   * @default ''
   * @property
   */
  headline?: string;
  /**
   * The description of the card
   * @default ''
   * @property
   */
  description?: string;
  /**
   * Image of the card
   * @default {}
   * @property
   */
  image?: ImageProps;
  /**
   * The icon Name for the card
   * @default ''
   * @property
   */
  icon?: string;
  /**
   * The size of the card supports 1 and 2,
   * based on the screen size, that will auto change the respective size of the card inbetween them
   * @default auto
   * @property
   */
  size?: 'auto' | '1' | '2';
  /**
   * onClick event for the card
   * @default undefined
   * @property
   */
  onClick?: () => void;
  /**
   * Cursor style for the card
   * @default false
   * @property
   */
  cursor?: boolean;

  /**
   * class name for the card
   * @default ''
   * @property
   */
  className?: string;
}

/**
 * Card component that will render the card with text and/or image
 *
 * The content of the card can the following combinations, "headline, description, icon", "headline and image" and "image".
 *
 * @group Molecules
 * @component
 *
 * @param {TextAndImageProps} props - The card props
 * @returns {React.ReactElement}
 *
 */
export const TextAndImage = ({
  variant,
  headline,
  description,
  icon,
  image,
  size,
  onClick,
  className,
  cursor,
}: TextAndImageProps) => {
  const id = useId();
  const cursorStyle = cursor ? styles.cursor : '';
  const showImageOnly =
    !headline && !description && !icon ? styles.showImageOnly : '';
  const showImage = image?.src ? styles.showImage : '';
  const classes = classNames(
    styles.container,
    showImageOnly,
    showImage,
    variant,
    styles['size-' + size],
    cursorStyle,
    className
  );

  return (
    <Box className={classes} onClick={onClick}>
      <Card
        className={styles.card}
        variant={variant === 'surface' ? variant : 'classic'}
        asChild
      >
        <React.Fragment key={id}>
          <Flex
            display="flex"
            direction="column"
            className={styles.cardContent}
            maxWidth="100%"
            height="100%"
          >
            {(headline || description) && (
              <Truncate className={styles.textSection}>
                <>
                  {headline && (
                    <Heading
                      align="left"
                      weight="semibold"
                      className={styles.heading}
                      size={{ initial: '2', md: '7' }}
                    >
                      {headline}
                    </Heading>
                  )}
                  {description && (
                    <Text
                      align="left"
                      as="div"
                      className={styles.description}
                      size={{ initial: '1', md: '4' }}
                    >
                      {description}
                    </Text>
                  )}
                </>
              </Truncate>
            )}
            {icon && (
              <Box className={styles.iconSection} asChild>
                <Icon name={icon} className={styles.icon} />
              </Box>
            )}
            {image && (
              <Image
                src={image.src}
                alt={image.alt ?? 'Card Image'}
                fallbackSrc={FallBackImg}
                className={styles.image}
              />
            )}
          </Flex>
        </React.Fragment>
      </Card>
    </Box>
  );
};

export default TextAndImage;
