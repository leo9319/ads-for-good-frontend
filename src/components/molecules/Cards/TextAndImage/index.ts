export { TextAndImage } from './TextAndImage';
export { default } from './TextAndImage';
export type { TextAndImageProps } from './TextAndImage';
