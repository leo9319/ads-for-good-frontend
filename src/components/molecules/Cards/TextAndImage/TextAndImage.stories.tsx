import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import TextAndImage from './TextAndImage';
import { Theme } from '@radix-ui/themes';

const meta = {
  title: 'Library/Molecules/Cards/TextAndImage',
  component: TextAndImage,
  tags: ['autodocs'],
  decorators: [
    Story => (
      <Theme
        style={{
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
          textAlign: 'center',
          backgroundColor: '#f7f7f7',
        }}
      >
        <Story />
      </Theme>
    ),
  ],
  argTypes: {
    variant: {
      control: {
        type: 'select',
      },
      options: ['deafult', 'surface'],
    },
    size: {
      control: {
        type: 'select',
      },
      options: ['auto', '1', '2'],
    },
    headline: {
      control: {
        type: 'text',
      },
    },
    description: {
      control: {
        type: 'text',
      },
    },
    icon: {
      control: {
        type: 'text',
      },
    },
    image: {
      control: {
        type: 'object',
      },
    },
  },
  args: {
    variant: 'default',
    size: 'auto',
  },
} satisfies Meta<typeof TextAndImage>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Text: Story = {
  args: {
    headline: 'Title Max 5 Lines',
    description: 'Description copy of no more than 5 lines',
    icon: 'icon-book',
  },
};

export const TextPlusImage: Story = {
  args: {
    headline: 'Title Max 5 Lines',
    description: 'Description copy of no more than 5 lines',
    image: {
      src: 'https://s3-alpha-sig.figma.com/img/0509/f8d9/3ea4462d6ad870c4dbf84df9cbcc4c34?Expires=1743379200&Key-Pair-Id=APKAQ4GOSFWCW27IBOMQ&Signature=p24CVG1watK-yODauTuzzRpz~ZxVBfirQCE0g-AQ-k5iMIJu5yeaQtHBzf13lJ-wErpAdyKqfutdWQTEH53TiQDJlr7bHFR2g-EqIWKYbi~GFH3qTzI3FnmJtRubW6MV6VgDjItnuPNy~~Ja88hjDJ1T9hlTqi0fAKJknWp8My1-6b43pRFL6wYPe55Qd~TMTupF~jcaW7UOf775cP9lHx-KwYk7-SadUt3q9OEbY0dWE86n0GznDkc3RnRDJ0dpHm3NupNAUHn53paDykZRABBJHHVF9o6EoJ2XAZIaJugBuup3b5SlyCP7-jkbCS1ZhEx7gT5zwqxeD0Yc2LZ~fg__',
      alt: 'alternate image text - Partial Image',
    },
  },
};

export const Image: Story = {
  args: {
    image: {
      src: 'https://s3-alpha-sig.figma.com/img/0509/f8d9/3ea4462d6ad870c4dbf84df9cbcc4c34?Expires=1743379200&Key-Pair-Id=APKAQ4GOSFWCW27IBOMQ&Signature=p24CVG1watK-yODauTuzzRpz~ZxVBfirQCE0g-AQ-k5iMIJu5yeaQtHBzf13lJ-wErpAdyKqfutdWQTEH53TiQDJlr7bHFR2g-EqIWKYbi~GFH3qTzI3FnmJtRubW6MV6VgDjItnuPNy~~Ja88hjDJ1T9hlTqi0fAKJknWp8My1-6b43pRFL6wYPe55Qd~TMTupF~jcaW7UOf775cP9lHx-KwYk7-SadUt3q9OEbY0dWE86n0GznDkc3RnRDJ0dpHm3NupNAUHn53paDykZRABBJHHVF9o6EoJ2XAZIaJugBuup3b5SlyCP7-jkbCS1ZhEx7gT5zwqxeD0Yc2LZ~fg__',
      alt: 'alternate image text 0 - Full Image',
    },
  },
};
