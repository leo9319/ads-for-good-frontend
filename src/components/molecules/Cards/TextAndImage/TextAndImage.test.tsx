import React from 'react';
import TextAndImage from './TextAndImage';
import { render } from '@testing-library/react';

describe('Truncate', () => {
  it('should render Text & Icon successfully', () => {
    const { baseElement } = render(
      <TextAndImage
        headline="Title Max 5 Lines"
        description="Description copy of no more than 5 lines"
        icon="icon-book"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render Text & Image successfully', () => {
    const { baseElement } = render(
      <TextAndImage
        headline="Title Max 5 Lines"
        description="Description copy of no more than 5 lines"
        icon="icon-book"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render Image successfully', () => {
    const { baseElement } = render(
      <TextAndImage
        headline="Title Max 5 Lines"
        description="Description copy of no more than 5 lines"
        icon="icon-book"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
