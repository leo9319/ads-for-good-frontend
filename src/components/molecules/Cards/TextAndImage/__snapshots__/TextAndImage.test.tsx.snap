// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Truncate should render Image successfully 1`] = `
<body>
  <div>
    <div
      class="rt-Box container size-undefined"
    >
      <div
        class="rt-Flex rt-r-display-flex rt-r-fd-column rt-r-max-w rt-r-h cardContent"
        style="--max-width: 100%; --height: 100%;"
      >
        <div
          class="rt-Box textSection container"
          style="--truncateBackgroundColor: white; --truncateLineLimit: 0;"
        >
          <div
            class="rt-Box truncateContent"
          >
            <h1
              class="rt-Heading rt-r-size-6 rt-r-ta-left headingContainer mocked-class-name typography-font-family-default heading"
              id=""
            >
              Title Max 5 Lines
            </h1>
            <div
              class="rt-Text rt-r-ta-left mocked-class-name typography-font-family-default description"
              id=""
            >
              Description copy of no more than 5 lines
            </div>
          </div>
        </div>
        <span
          class="rt-Box iconSection icon"
          style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
        >
          <span
            class="icon icon-book"
          />
        </span>
      </div>
    </div>
  </div>
</body>
`;

exports[`Truncate should render Text & Icon successfully 1`] = `
<body>
  <div>
    <div
      class="rt-Box container size-undefined"
    >
      <div
        class="rt-Flex rt-r-display-flex rt-r-fd-column rt-r-max-w rt-r-h cardContent"
        style="--max-width: 100%; --height: 100%;"
      >
        <div
          class="rt-Box textSection container"
          style="--truncateBackgroundColor: white; --truncateLineLimit: 0;"
        >
          <div
            class="rt-Box truncateContent"
          >
            <h1
              class="rt-Heading rt-r-size-6 rt-r-ta-left headingContainer mocked-class-name typography-font-family-default heading"
              id=""
            >
              Title Max 5 Lines
            </h1>
            <div
              class="rt-Text rt-r-ta-left mocked-class-name typography-font-family-default description"
              id=""
            >
              Description copy of no more than 5 lines
            </div>
          </div>
        </div>
        <span
          class="rt-Box iconSection icon"
          style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
        >
          <span
            class="icon icon-book"
          />
        </span>
      </div>
    </div>
  </div>
</body>
`;

exports[`Truncate should render Text & Image successfully 1`] = `
<body>
  <div>
    <div
      class="rt-Box container size-undefined"
    >
      <div
        class="rt-Flex rt-r-display-flex rt-r-fd-column rt-r-max-w rt-r-h cardContent"
        style="--max-width: 100%; --height: 100%;"
      >
        <div
          class="rt-Box textSection container"
          style="--truncateBackgroundColor: white; --truncateLineLimit: 0;"
        >
          <div
            class="rt-Box truncateContent"
          >
            <h1
              class="rt-Heading rt-r-size-6 rt-r-ta-left headingContainer mocked-class-name typography-font-family-default heading"
              id=""
            >
              Title Max 5 Lines
            </h1>
            <div
              class="rt-Text rt-r-ta-left mocked-class-name typography-font-family-default description"
              id=""
            >
              Description copy of no more than 5 lines
            </div>
          </div>
        </div>
        <span
          class="rt-Box iconSection icon"
          style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
        >
          <span
            class="icon icon-book"
          />
        </span>
      </div>
    </div>
  </div>
</body>
`;
